README
*** System Installer Project ***

SystemInstaller is a tool to install a system with any Linux distribution. 
It works hand-in-hand with SystemImager and SystemConfigurator to build 
a cluster.

***********************************************************************

Since version 1.04 the OSCAR project was the only active user and maintainer of
SystemInstaller. Several extensions were added to version 1.04 and maintained
as patches inside the OSCAR SRPM file. At the beginning of the year 2006 the OSCAR
specific version of SystemInstaller was finally checked into the OSCAR group's
SVN repository in order to facilitate development and the addition of OSCAR
specific features.

THIS BRANCH OF SYSTEMIMAGER MAKES SENSE ONLY WHEN USED WITH OSCAR!

The integration with OSCAR can onlz be improved if unnecessary components
of SystemImager don't need to be maintained any more. This version will be
systematically stripped by not needed and obsolete components.

Erich Focht, February 12, 2006.

************************************************************************

The original SystemInstaller README file is included below. Some info is
obsolete.

************************************************************************


If you wish to help out in any way, please visit the 
SystemInstaller home page (http://systeminstaller.sourceforge.net).

*** NOTE ***

System Installer is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License located in the COPYING file of this
distribution for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

***************************************************************
        Overview and quick tips
***************************************************************

For more detailed info on SIS in general, including step by step
instructions, see the SIS manpage.

SystemInstaller is a tool to build Linux OS images for use
by SystemImager. It is designed to be distributions and
architecture independent, although we aren't 100% there yet.
After installing the systeminstaller package and optionally, 
the systeminstaller-x11 package, you can get started.

The easiest (and prettiest) way to go about your installation
is to use the SystemInstaller gui, tksis which is in the 
systeminstaller-x11 package (you did install that, didn't you?)
Once you have launched tksis, you can use the panels there to add
new images and clients. 

Optionally there is a text based interactive interface for defining
the necessary components. It is called buildimage and can be run
directly from the command line. It will ask a bunch of questions,
providing defaults where it makes sense. After buildimage completes
you will have an image and clients defined.

There is yet a third option. There is a command line interface for
all the SystemInstaller functions. Briefly, these commands are:

	-mksiimage	Creates a Linux OS image
	-mksidisk	Creates a disk partition table
	-mksirange	Creates a range of machines
	-mksimachine	Edits machine definitions
	-mkdhcpconf	Creates a dhcpd.conf file

See the man pages for these commands for more detail.

In order to actually install the clients, you will need to either
make an autoinstall diskette or CD or setup network booting. See
the SystemImager documentation for more details on this setup.

***************************************************************
        Prerequisites
***************************************************************
SystemInstaller is part of the System Installation Suite. As
a result it has several prerequisites on other components of 
SIS(SystemImager and System Configurator). Links to the 
project pages and hence the downloads can be found at 
sisuite.org.

***************************************************************
        Getting and giving help
***************************************************************

The SystemInstaller project is hosted on SourceForge. You can 
view our web page at http://systeminstaller.sourceforge.net/.

We have mailing lists for both users and developers. They can
be reached at http://sourceforge.net/mail/?group_id=24005

Please don't hesitate to contact the lists with any questions,
comments, or suggestions.

***************************************************************
	A note about Debian support
***************************************************************

Debian support was added to SystemInstaller recently, but since
we can't seem to get around the interactive questions that it insists
on asking. So it hasn't been actively maintained, most notably,
it hasn't been modded to support the new API spec. Hopefully the
debconf in the next version of Debian will help with this situation.
If you want to still use the Debian stuff, let us know and we'll 
port it over to the new API.

Here is the original text that was under this item.

Support for .deb packages has recently been added to SystemInstaller,
but only at a very minimal level.  Systeminstaller can create an image
with the base system and any .deb packages which do not require
interactive setup.  This means you must provide in your package list
the base system tarball (base2_2.tgz for debian/stable).

There is a solution to the packages that require interactive setup,
and it is coming RSN.  The short term solution will force the
interactive setup to use the defaults, and the
less-short-term-but-not-far-away solution will allow for you to answer
the questions as they are prompted by the debian setup.


CHANGELOG
------------------------------------------------------------
System Installer v2.6.6
------------------------------------------------------------
- Fix for mksiimage so it won't destroy host binded filesystems
  if bootstrap fails.

------------------------------------------------------------
System Installer v2.6.5
------------------------------------------------------------
- Fix for mksiimage that was broken (systemimager.json changes)

------------------------------------------------------------
System Installer v2.6.4
------------------------------------------------------------
- Fix for tksis that was broken long long ago (SIS DB change)

------------------------------------------------------------
System Installer v2.6.3
------------------------------------------------------------
- Migrated to new SystemImager::JConfig v4.9.0-0.20191211git

------------------------------------------------------------
System Installer v2.6.2
------------------------------------------------------------
- Migrated to new SystemImager v4.9.0

------------------------------------------------------------
System Installer v2.6.1
------------------------------------------------------------
- Disabled si_mkautoinstallscript as it's already done by
  OSCAR::ImageMgt::create_image().

------------------------------------------------------------
System Installer v2.6.0
------------------------------------------------------------
- Now mksiimage can create non-oscarimage
- Added --coreopkgs to add core opkgs to the image
- Added --selectedopkgs to add selected + core opkgs to the image.

------------------------------------------------------------
System Installer v2.5.2
------------------------------------------------------------
- Optimized mksiimage
- Moved mount/unmount code to OSCAR::PackMan.

------------------------------------------------------------
System Installer v2.5.1
------------------------------------------------------------
- Bugfix release.

------------------------------------------------------------
System Installer v2.5.0
------------------------------------------------------------
- Add support for choosing root filesystem for client in the GUI.

------------------------------------------------------------
System Installer v2.4.18
------------------------------------------------------------
- Bugfix (badly generated config file (HASH instead of value)

------------------------------------------------------------
System Installer v2.4.17
------------------------------------------------------------
- New version that adds support for options:
  --imageserver, --logserverport, --sshdownloadurl

------------------------------------------------------------
System Installer v2.4.16
------------------------------------------------------------
    - Fix missing option in dhcpd.conf
      need to add: "use-host-decl-names on;" otherwise
      hostname is not seen by dhclient.

------------------------------------------------------------
System Installer v2.4.15
------------------------------------------------------------
    - Fix padding bug.

------------------------------------------------------------
System Installer v2.4.14
------------------------------------------------------------
    - Bugfix release.

------------------------------------------------------------
System Installer v2.4.13
------------------------------------------------------------
    - Bugfix release due to API change in prepare_pools

------------------------------------------------------------
System Installer v2.4.12
------------------------------------------------------------
    - Add missing nodes infos required by oscar and add GPU support.
    - Make some error messages more explicit.
    - Fix issues with newer version of DHCP (patch from Olivier
      Lahaye for the RHEL6 support).
    - Add the support for initramfs (path from Olivier Lahaye
      for the RHEL6 support).
    - Fix the detection of the kernel on CentOS 6.
    - Move to INSTALLDIRS=vendor

------------------------------------------------------------
System Installer v2.4.11
------------------------------------------------------------
    - Do not try to setup DNS within the DHCP configuration on a private
      network (patch from Olivier Lahaye).
    - Improve the management of the /etc/hosts when adding new clients.
    - Improve the distro detection when creating a new image.

------------------------------------------------------------
System Installer v2.4.10 
------------------------------------------------------------
    - Clean up the code for the management of partitions within an image.
    - Fix a few issues when dealing with images' partitions.

------------------------------------------------------------
System Installer v2.4.9
------------------------------------------------------------
    - Improve the help output of the mksiimage command (patch from Olivier 
      Lahaye).
    - Continue to fix issues with CCISS controllers (patch from Olivier
      Lahaye).
    - Improve the disk partitioning capability, e.g., support of GTP 
      partitions (patch from Olivier Lahaye).

------------------------------------------------------------
System Installer v2.4.8 
------------------------------------------------------------
    - Fix a bug in the Partition/IA.pm.

------------------------------------------------------------
System Installer v2.4.7
------------------------------------------------------------
    - Remove debugging info.
    - Improve the support of CCISS controllers.
    - First set of modifications to support again client removal.

------------------------------------------------------------
System Installer v2.4.6
------------------------------------------------------------
    - More asserts to try to avoid segfaults.
    - Improve command output.
    - Fix partitionning using when using CCISS controllers 
      (patch from Olivier Lahaye).

------------------------------------------------------------

lib/SystemInstaller/Env.pm
package SystemInstaller::Env;

#   Copyright (c) 2007 Erich Focht <efocht@hpce.nec.com>
#                      All rights reserved.
 
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
 
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
 
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 
#   Michael Chase-Salerno <salernom@us.ibm.com>
use strict;
use warnings;
use 5.010;

BEGIN {
    use Exporter();

    @SystemInstaller::Env::ISA       = qw(Exporter);
    @SystemInstaller::Env::EXPORT    = qw(print_version get_version);
    @SystemInstaller::Env::EXPORT_OK = qw($config);

}

use vars qw(@EXPORT $config);
use AppConfig;
use SystemImager::JConfig qw($jconfig);

# Globally needed variables
$config = AppConfig->new(
        cfgfile =>      { ARGCOUNT => 1, 
                DEFAULT => '/etc/systeminstaller/systeminstaller.conf'},
        binpath =>      { ARGCOUNT => 1, 
                DEFAULT => '/usr/local/sbin:/usr/bin:/usr/sbin'},
        pkgcachefile => { ARGCOUNT => 1, DEFAULT => '.pkgcache'},
        simachine =>    { ARGCOUNT => 1, DEFAULT => 'mksimachine'},
        sirange =>      { ARGCOUNT => 1, DEFAULT => 'mksirange'},
        siimage =>      { ARGCOUNT => 1, DEFAULT => 'mksiimage'},
        sidisk =>       { ARGCOUNT => 1, DEFAULT => 'mksidisk'},
        mkaiscript =>   { ARGCOUNT => 1, DEFAULT => 'si_mkautoinstallscript'},
        addclients =>   { ARGCOUNT => 1, DEFAULT => 'si_addclients'},
        delimage =>     { ARGCOUNT => 1, DEFAULT => 'si_rmimage'},
        cpimage =>      { ARGCOUNT => 1, DEFAULT => 'si_cpimage'},
        distinfo =>     { ARGCOUNT => 1, 
                DEFAULT => '/usr/share/systeminstaller/distinfo'},
        pkgpath =>      { ARGCOUNT => 1, DEFAULT => '/var/lib//tftpboot/rpms'},
        ipmeth =>       { ARGCOUNT => 1, DEFAULT => 'dynamic_dhcp'},
        piaction =>     { ARGCOUNT => 1, DEFAULT => 'beep'},
        disktype =>     { ARGCOUNT => 1, DEFAULT => 'scsi'},
        rpmrc =>        { ARGCOUNT => 1, DEFAULT => '/usr/lib/rpm/rpmrc'},
        rpm =>          { ARGCOUNT => 1, DEFAULT => 'rpm'},
        rpmargs =>      { ARGCOUNT => 1, DEFAULT => "-vh"},
        dpkg =>         { ARGCOUNT => 1, DEFAULT => 'dpkg'},
        verbose =>      { ARGCOUNT => 0, ALIAS => "v"},
        pkginstfail =>  { ARGCOUNT => 0},
        postinstfail => { ARGCOUNT => 0},
        diskversion =>  {ARGCOUNT => 1, DEFAULT => "2"},
        # The next are variables that we expect from the SystemImager
        # systemimager.conf file.
        default_image_dir         => { ARGCOUNT => 1 },
        default_override_dir      => { ARGCOUNT => 1 },
        autoinstall_script_dir    => { ARGCOUNT => 1 },
        autoinstall_boot_dir      => { ARGCOUNT => 1 },
        autoinstall_tarball_dir   => { ARGCOUNT => 1 },
        autoinstall_torrent_dir   => { ARGCOUNT => 1 },
        rsyncd_conf               => { ARGCOUNT => 1 },
        rsync_stub_dir            => { ARGCOUNT => 1 },
        tftp_dir                  => { ARGCOUNT => 1 },
        net_boot_default          => { ARGCOUNT => 1 },
        # The next are variables that we expect from tksis.conf
        icon_dir => { ARGCOUNT => 1, DEFAULT => "/usr/share/systeminstaller/images"},
        xterm_cmd => { ARGCOUNT => 1,
            DEFAULT => "xterm -bg black -fg magenta",
        },
);

if (-e $config->cfgfile ) {
        $config->file($config->cfgfile);
}

if (-e '/etc/systeminstaller/tksis.conf') {
    $config->file('/etc/systeminstaller/tksis.conf');
}

#if (-e '/etc/systemimager/systemimager.conf') {
#    $config->file('/etc/systemimager/systemimager.conf');
#}
# systemimager.conf is deprecated. use $jconfig to insert needed values in $config.
$config->set('default_image_dir', $main::jconfig->get('imager','images_dir'));
$config->set('default_override_dir', $main::jconfig->get('imager','overrides_dir'));
$config->set('autoinstall_script_dir', $main::jconfig->get('imager','scripts_dir'));
$config->set('autoinstall_boot_dir', $main::jconfig->get('pxe','boot_files'));
$config->set('autoinstall_tarball_dir', $main::jconfig->get('xmit_torrent','tarballs_dir'));
$config->set('autoinstall_torrent_dir', $main::jconfig->get('xmit_torrent','torrents_dir'));
$config->set('rsyncd_conf', $main::jconfig->get('xmit_rsync','config_file'));
$config->set('rsync_stub_dir', $main::jconfig->get('xmit_rsync','stubs_dir'));
$config->set('tftp_dir', $main::jconfig->get('pxe','tftp_dir'));
$config->set('net_boot_default', $main::jconfig->get('pxe','boot_mode'));


# Push it up to main
$::main::config = $config;

sub print_version {
        my $CMD=shift;
        $CMD =~ s/^.*\///;
        my $CMDVERSION=shift;
        my $PKGVERSION=&get_version;
        print <<EOL;
$CMD version $CMDVERSION
Part of SystemInstaller version $PKGVERSION

Copyright (C) 2001 Internation Business Machines
Copyright (C) 2007 Erich Focht @ NEC HPC Europe
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

EOL
            return 1;
}

sub get_version {
        # This just returns the version number, looks silly,
        # but the string below is replaced during the build
        # process with the proper version.
        my $SIVERSION="2.6.4";
        return $SIVERSION;
}


### POD from here down

=head1 NAME
 
SystemInstaller::Env - Environment info
 
=head1 SYNOPSIS   

 use SystemInstaller::Env;

 print_version("mksidisk","1.2.2");

=head1 DESCRIPTION

SystemInstaller::Env defines on object containing environment info for SystemInstaller.

It exports an Appconfig structure, $config into the "main" name space.

Also provided is the print_version subroutine that will print the version blurb for
the current version of SystemInstaller. It takes the command name and its file version
as input and always returns 1.

=head1 AUTHOR
 
Michael Chase-Salerno <mchasal@users.sourceforge.net>

=head1 SEE ALSO

L<AppConfig>, L<perl>, L<systeminstaller.conf>

 
=cut

1;
